pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+		return ua.attachToLibrariesOfPID(pid)
+	}
+
+	return nil
+}
+
+// DetachPID detaches the uprobes attached to a PID
+func (ua *UprobeAttacher) DetachPID(pid uint32) error {
+	return ua.fileRegistry.Unregister(pid)
+}
+
+const buildKitProcessName = "buildkitd"
+
+func isBuildKit(procInfo *ProcInfo) bool {
+	comm, err := procInfo.Comm()
+	if err != nil {
+		return false
+	}
+	return strings.HasPrefix(comm, buildKitProcessName)
+}
+
+func isContainerdTmpMount(path string) bool {
+	return strings.Contains(path, "tmpmounts/containerd-mount")
+}
+
+// getUID() return a key of length 5 as the kernel uprobe registration path is limited to a length of 64
+// ebpf-manager/utils.go:GenerateEventName() MaxEventNameLen = 64
+// MAX_EVENT_NAME_LEN (linux/kernel/trace/trace.h)
+//
+// Length 5 is arbitrary value as the full string of the eventName format is
+//
+//	fmt.Sprintf("%s_%.*s_%s_%s", probeType, maxFuncNameLen, functionName, UID, attachPIDstr)
+//
+// functionName is variable but with a minimum guarantee of 10 chars
+func getUID(lib utils.PathIdentifier) string {
+	return lib.Key()[:5]
+}
+
+func parseSymbolFromEBPFProbeName(probeName string) (symbol string, isManualReturn bool, err error) {
+	parts := strings.Split(probeName, "__")
+	if len(parts) < 2 {
+		err = fmt.Errorf("invalid probe name %s, no double underscore (__) separating probe type and function name", probeName)
+		return
+	}
+
+	symbol = parts[1]
+	if len(parts) > 2 {
+		if parts[2] == "return" {
+			isManualReturn = true
+		} else {
+			err = fmt.Errorf("invalid probe name %s, unexpected third part %s. Format should be probeType__funcName[__return]", probeName, parts[2])
+			return
+		}
+	}
+
+	return
+}
+
+// attachToBinary attaches the probes to the given binary. Important: it does not perform any cleanup on failure.
+// This is to match the behavior of the FileRegistry, which will call the deactivation callback on failure of the registration
+// callback.
+func (ua *UprobeAttacher) attachToBinary(fpath utils.FilePath, matchingRules []*AttachRule, procInfo *ProcInfo) error {
+	if ua.config.ExcludeTargets&ExcludeBuildkit != 0 && isBuildKit(procInfo) {
+		return fmt.Errorf("process %d is buildkitd, skipping", fpath.PID)
+	} else if ua.config.ExcludeTargets&ExcludeContainerdTmp != 0 && isContainerdTmpMount(fpath.HostPath) {
+		return fmt.Errorf("path %s from process %d is tempmount of containerd, skipping", fpath.HostPath, fpath.PID)
+	}
+
+	symbolsToRequest, err := ua.computeSymbolsToRequest(matchingRules)
+	if err != nil {
+		return fmt.Errorf("error computing symbols to request for rules %+v: %w", matchingRules, err)
+	}
+
+	inspectResult, isAttachable, err := ua.inspector.Inspect(fpath, symbolsToRequest)
+	if err != nil {
+		return fmt.Errorf("error inspecting %s: %w", fpath.HostPath, err)
+	}
+	if !isAttachable {
+		return fmt.Errorf("incompatible binary %s", fpath.HostPath)
+	}
+
+	uid := getUID(fpath.ID)
+
+	for _, rule := range matchingRules {
```

Comment:
please break this 4 nested loops into functions

Reply:
also, we can leverage the iterator pattern

Reply:
I moved the two innermost loops to a separate function, although I'm not sure what you mean by the iterator pattern.